Hot rank update batching + deadlock avoidance

sunaurus opened the pull request:
This PR contains the following changes:
1. **Ensure that comment_aggregates table rows are always updated in the same order to help with deadlocks** 
2. **Reduce frequency of hot_rank updates from every 5 minutes to every 15 minutes**
    * If this is controversial, then I will happily revert it, but my reasoning was that it would give the hot rank scheduled task much more breathing room in case the amount of comments or posts on Lemmy increases exponentially in the near future. At the same time, 15 minutes should still be regular enough to keep the sorting updates feel quite alive.
3. **Instead of doing hot rank updates in one big UPDATE statement which can lock the whole table for 10+ seconds, the updates are now done in batches of 5000 rows**
    * This is done with the help of a new `hot_rank_updated` column - we can use this to ensure that we process every row only once per each scheduled task run.
    * I tried a few different batch sizes, including 100, 1000, 5000 and 10 000. On my instance, we have ~170k rows in `comment_aggregates` and ~28k rows in `post_aggregates` for the last week. Updating hot ranks for all of them with batches of 5000 takes <30 seconds total. For comparison, with batches of 100, it took about 4 minutes. (It could be that on different hardware, different batch sizes may work better)

-----

I have been running this code (cherry-picked onto 0.17.4) live on lemm.ee for the past hour (starting from 05:15 on the graph below), and I have not seen a single deadlock since deploying these changes:

![image](https://github.com/LemmyNet/lemmy/assets/5356547/27b00650-1c5e-4aa5-9adc-4dcf9dbd2efc)


----
This should fix #3076. 

Thanks to [@phiresky](https://github.com/phiresky) for your input about all this [here](https://github.com/LemmyNet/lemmy/issues/3076#issuecomment-1594881225)

sunaurus commented on src/scheduled_tasks.rs:
```diff
@@ -70,66 +70,127 @@ pub fn setup(db_url: String, user_agent: String) -> Result<(), LemmyError> {
+  loop {
+    let mut post_select = post_aggregates::table
+      .select(post_aggregates::post_id)
+      .filter(post_aggregates::hot_rank_updated.lt(update_start_time))
```
Hmm, but in this case, if there is some hugely popular post (or even comment thread) within some time bucket, then the load will be spread quite unevenly and we can still end up with an UPDATE statement that can have several tens (or why not even hundreds) of thousands of touched rows. Or am I misunderstanding you?

dessalines replied:
I'd say even with that, it'd still be *good enough* for batching, especially since that number is tunable and simple. As long as there's enough buckets we should be okay. IE 7, 1-day buckets might even be enough to prevent the deadlocks, but in code we could make that `1000` buckets.

The other alternative is doing a select to gather up all the IDs for the last 7 days, then doing an `eq_any` / in query, although those are generally extremely inefficient and wasteful.

phiresky replied:
You could also do it without bucketing by doing "normal" timestamp pagination: (pseudocode)

last_timestamp = '1970-01-01' // or last week i guess
do {
rows = select comment_id from comments where published > last_timestamp order by published asc limit batch_size;
last_timestamp = rows[-1].published;
rows = rows.sortby(comment_id) // needs to be consistent with other multi-row update queries for deadlocks
} while len(rows) == batch_size

* batches have fixed sizes
* no parallelization possible because next batch is only determined after current batch
* need to look for off-by-one error if two rows have the same timestamp


advantage of the last_update column is that you know which rows still need updating but yeah seems like it might not be needed at this point

sunaurus replied:
Yeah, I'm not sure why I didn't go with that approach to begin with - I remember considering it and then thinking it would potentially never finish processing in some cases, but I think I was just a bit sleep deprived and confused myself 😄 

@dessalines would you be OK with that approach as well?

dessalines replied:
I'm not exactly sure how that would translate to SQL / diesel, and it seems a bit more complicated than just creating a vector of equally-spaced times with a *between* query, but sure. As long as it doesn't use an `eq_any` / in query, it should be fine.

phiresky commented:
Great! But I think you have an issue with your query performance because you're filtering by a different column than you're ordering by. In order for your selects to be performant you have to filter and order by the same column and that column must be indexed. 

To verify, try running

1. `explain analyze select comment_id from comment_aggregate where hot_rank_update < '2024-01-01' order by comment_id asc limit 10`. this should do a slow scan via the comment_id index, filtering out rows one-by-one. it will get slower and slower the more comments are already updated because it always scans from the left of ids
2. make sure you have an `create index on comment_aggregate(hot_rank_update);`. then run `explain analyze select comment_id from comment_aggregate where hot_rank_update < '2024-01-01' order by hot_rank_update asc limit 10`. this should do an index scan on hot_rank_update and always only look at 10 rows total. even better, leave the order_by out completely and just let pg return the first rows it finds however it wants.

I know I said you have to do deterministic ordering to fix deadlocks. But you can still do that after selecting, they only have to be ordered by comment_id _within each bach_.

(I don't have an instance to verify what i'm saying, just looking at the code)

sunaurus commented:
I pushed a new version. 

* Performance is improved now compared to original + no additional db columns are used.
* Batches are now constructed based on the existing `published` column
* Each batch is selected and updated in a single DB query
* Deadlocks are mitigated with `SELECT ... FOR UPDATE SKIP LOCKED`
* Batch size reduced to 1000, as bigger batches were not using indexes
* Added indexes on `published` for `comment_aggregates` and `community_aggregates` (`post_aggregates` already had it)

The main benefit of doing it with a fixed batch size is that this approach will scale linearly with amount of comments, which I believe might be critically important for Lemmy going forward.

Caveat: I unfortunately had to use `sql_query` for this due to Diesel not supporting what we need here. I also wrote an alternate version with Diesel, using two separate queries in a transaction, and despite that version having roughly the same query plan, it performed about 4-5x worse on average than this `sql_query` based version (presumably due to serialization between the two queries). So it seems the trade-off of using `sql_query` is worth it in this case.

Btw, thanks again to @phiresky for additional help!

dessalines commented on src/scheduled_tasks.rs:
```diff
@@ -100,64 +93,93 @@ fn startup_jobs(db_url: &str) {
+      r#"WITH batch AS (SELECT a.id
+               FROM {aggregates_table} a
+               WHERE a.published > $1
+               ORDER BY a.published
```
The indexes you added were `published desc`, would be good to use that. Or if that doesn't work because of the incrementing `previous_batch_last_published` used here, then make those indexes `asc`

sunaurus replied:
I went with `desc` indexes for the two new ones, because the existing one on `post_aggregates` already had a `desc` index, so I thought it looked more uniform 😃

AFAIK there should be no performance difference between forward and backward index scans, do you know differently?

Btw, this query is definitely using the index:

```
Update on post_aggregates a  (cost=83.81..2078.90 rows=1000 width=38) (actual time=0.300..172.142 rows=1000 loops=1)
  CTE batch
    ->  Limit  (cost=0.42..83.40 rows=1000 width=88) (actual time=0.062..8.010 rows=1000 loops=1)
          ->  LockRows  (cost=0.42..2630.48 rows=31697 width=88) (actual time=0.060..7.637 rows=1000 loops=1)
                ->  Index Scan Backward using idx_post_aggregates_published on post_aggregates a_1  (cost=0.42..2313.51 rows=31697 width=88) (actual time=0.044..5.588 rows=1000 loops=1)
                      Index Cond: (published > (now() - '7 days'::interval))
  ->  Nested Loop  (cost=0.42..1995.50 rows=1000 width=38) (actual time=0.256..138.719 rows=1000 loops=1)
        ->  CTE Scan on batch  (cost=0.00..20.00 rows=1000 width=32) (actual time=0.077..11.958 rows=1000 loops=1)
        ->  Index Scan using post_aggregates_pkey on post_aggregates a  (cost=0.42..1.72 rows=1 width=26) (actual time=0.007..0.007 rows=1 loops=1000)
              Index Cond: (id = batch.id)
Planning Time: 0.463 ms
Execution Time: 172.697 ms
```

phiresky replied:
DESC and ASC indexes can both be used for either direction of query, they have a slight performance difference that's afaik only really relevant when the index is not in RAM and you're scanning through a large amount of rows in the index (it determines the ordering on disk where fetching forward is more efficient than fetching backwards due to prefetch)

dessalines commented on src/scheduled_tasks.rs:
```diff
@@ -100,64 +93,93 @@ fn startup_jobs(db_url: &str) {
+) {
+  let update_batch_size = 1000; // Bigger batches than this tend to cause seq scans
+  let mut previous_batch_last_published = process_start_time;
+  loop {
```
Something tells me there's a better way to do this than using loop, maybe something like `while` or `while let`, but your match seems to check the conditions correctly anyway.

sunaurus replied:
You're definitely right, `while let` makes it more clear and readable, I changed it 👍

sunaurus commented on src/scheduled_tasks.rs:
```diff
@@ -100,64 +93,93 @@ fn startup_jobs(db_url: &str) {
+               WHERE a.published > $1
+               ORDER BY a.published
+               LIMIT $2
+               FOR UPDATE SKIP LOCKED)
```
Diesel has `.for_update()` and `.skip_locked()`, but they have some significan't limitations, most notably they can't be used together with `.into_boxed()`

dessalines replied:
Gotcha. @Nutomic would these be potentially useful in any apub jobs?

Nutomic replied:
This seems to be useful when using a table as a kind of job queue? Might be useful for https://github.com/LemmyNet/lemmy/issues/2142 once we implement that.

dessalines reviewed (approved):
Thanks a ton for this, this is gonna work so much better than what's currently there for bigger instances.

Nutomic commented on src/scheduled_tasks.rs:
```diff
@@ -100,66 +93,92 @@ fn startup_jobs(db_url: &str) {
+      Ok(updated_rows) => previous_batch_result = updated_rows.last().map(|row| row.published),
+      Err(e) => {
+        error!("Failed to update {} hot_ranks: {}", table_name, e);
+        break;
```
Should it really stop processing new batches if any one batch threw an error? Seems unnecessary.
